doc/man/doveadm-pw.1.in
.\" Copyright (c) 2010-2018 Dovecot authors, see the included COPYING file
.TH DOVEADM\-PW 1 "2015-06-05" "Dovecot v2.3" "Dovecot"
.SH NAME
doveadm\-pw \- Dovecot\(aqs password hash generator
.\"------------------------------------------------------------------------
.SH SYNOPSIS
.BR doveadm " [" \-Dv "] " "pw \-l"
.\"-------------------------------------
.br
.BR doveadm " [" \-Dv "] " pw
[\fB\-p\fP \fIpassword\fP]
[\fB\-r\fP \fIrounds\fP]
[\fB\-s\fP \fIscheme\fP]
[\fB\-u\fP \fIuser\fP]
.RB [ \-V ]
.\"-------------------------------------
.br
.BR doveadm " [" \-Dv "] " "pw \-t"
.I hash
[\fB\-p\fP \fIpassword\fP]
[\fB\-u\fP \fIuser\fP]
.\"------------------------------------------------------------------------
.SH DESCRIPTION
.B doveadm pw
is used to generate password hashes for different password
.IR scheme s
and optionally verify the generated hash.
.PP
All generated password hashes have a
.RI { scheme }
prefix, for example
.RB { SHA512\-CRYPT.HEX }.
All passdbs have a default scheme for passwords stored without the
.RI { scheme }
prefix.
The default scheme can be overridden by storing the password with the
scheme prefix.
.PP
.\"------------------------------------------------------------------------
@INCLUDE:global-options@
.\" --- command specific options --- "/.
.PP
Command specific
.IR options :
.\"-------------------------------------
.TP
.B \-l
List all supported password
.IR scheme s
and exit successfully.
.br
There are up to three optional password
.IR scheme s:
.BR BLF\-CRYPT " (Blowfish crypt),"
.BR SHA256\-CRYPT\  and
.BR SHA512\-CRYPT .
Their availability depends on the system\(aqs currently used libc.
.\"-------------------------------------
.TP
.BI \-p\  password
The plain text
.I password
for which the hash should be generated.
If no
.I password
was given
.BR doveadm (1)
will prompt interactively for one.
.\"-------------------------------------
.TP
.BI \-r\  rounds
The password
.IR scheme s
.BR BLF\-CRYPT ,
.BR SHA256\-CRYPT\  and
.B SHA512\-CRYPT
supports a variable number of encryption
.IR rounds .
The following table shows the minimum/maximum number of encryption
.I rounds
per scheme.
When the
.B \-r
option was omitted the default number of encryption rounds will be applied.
.\"
.sp
.nf
 Scheme       | Minimum | Maximum   | Default
----------------------------------------------
 BLF\-CRYPT    |       4 |        31 |       5
 SHA256\-CRYPT |    1000 | 999999999 |    5000
 SHA512\-CRYPT |    1000 | 999999999 |    5000
.fi
.\"
.\"-------------------------------------
.TP
.BI \-s\  scheme
The password
.I scheme
which should be used to generate the hashed password.
By default the
.BI CRYPT\  scheme
will be used (with the $2y$ bcrypt format).
It is also possible to append an encoding suffix to the
.IR scheme .
Supported encoding suffixes are:
.BR .b64 ,
.BR .base64\  and
.BR .hex .
.br
See also
https://doc.dovecot.org/configuration_manual/authentication/password_schemes/
for more details about password schemes.
.\"-------------------------------------
.TP
.BI \-t\  hash
Test if the given password
.I hash
matches a given plain text password.
You should enclose the password
.I hash
in single quotes, if it contains one or more dollar signs
.RB ( $ ).
The plain text password may be passed using the
.BR \-p " option."
When no password was specified,
.BR doveadm (1)
will prompt interactively for one.
.\"-------------------------------------
.TP
.BI \-u\  user
When the
.BI DIGEST\-MD5\  scheme
is used, also the
.I user
name must be given, because the user name is a part of the generated hash.
For more information about Digest\-MD5 please read also:
https://doc.dovecot.org/configuration_manual/authentication/digest-md5/
.\"-------------------------------------
.TP
.B \-V
When this option is given, the hashed password will be internally verified.
The result of the verification will be shown after the hashed password,
enclosed in parenthesis.
.\"------------------------------------------------------------------------
.SH EXAMPLE
The first password hash is a DIGEST\-MD5 hash for jane.roe@example.com.
The second password hash is a CRAM\-MD5 hash for john.doe@example.com.
.sp
.nf
.B doveadm pw \-s digest\-md5 \-u jane.roe@example.com
Enter new password:
Retype new password:
{DIGEST\-MD5}9b9dcb4466233a9307bbc33708dffda0
.B doveadm pw
Enter new password:
Retype new password:
{CRAM\-MD5}913331d8782236a8ecba7764a63aa27b26437fd40ca878d887f11d81245c2c6b
.fi
.\"------------------------------------------------------------------------
@INCLUDE:reporting-bugs@
.\"------------------------------------------------------------------------
.SH SEE ALSO
.BR doveadm (1)

NEWS
v2.3.17.1 2021-12-07  Aki Tuomi <aki.tuomi@open-xchange.com>

	- dsync: Add back accidentically removed parameters.
	- lib-ssl-iostream: Fix assert-crash when OpenSSL returned syscall error
	  without errno.
	- master: Dovecot failed to start if ssl_ca was too large.

v2.3.17 2021-10-28  Aki Tuomi <aki.tuomi@open-xchange.com>

	* Dovecot now logs a warning if time seems to jump forward at least
	  100 milliseconds.
	* dict: Lines logged by the dict process now contain the dict name as
	  the prefix.
	* lib-index: mail_cache_fields, mail_always_cache_fields and
	  mail_never_cache_fields now verifies that the listed header names are
	  valid. Especially the UTF8 "–" character has sometimes been wrongly
	  used instead of the ASCII "-".
	+ *-login: Added login_proxy_rawlog_dir setting to capture
	  rawlogs between proxy and backend.
	+ dict: The server process now keeps the last 10 idle dict backends
	  cached for maximum of 30 seconds. Practically this acts as a
	  connection pool for dict-redis and dict-ldap. Note that this doesn't
	  affect dict-sql, because it already had its own internal cache.
	+ doveadm: New stats add/remove commands added to support changing the
	  metrics configuration on runtime.
	+ lazy_expunge: Added lazy_expunge_exclude settings to disable
	  lazy_expunge for specific folders. \Special-use flags can be used as
	  folder names.
	+ lib-lua: Added a new helper function dovecot.restrict_global_variables()
	  to disable or enable defining new global variables.
	- LAYOUT=index List index rebuild was missing.
	- LAYOUT=index: Duplicate GUIDs were not detected.
	- acl: When using acl_ignore_namespace Dovecot attempted to access or
	  create dovecot-acl-list even when the namespace should have been
	  ignored. For virtual namespaces this could have yielded errors about
	  "Read-only file system" or "Permission denied".
	- auth: Setting the "master" passdb field to empty value would
	  cause proxying to fail with an authentication error.
	  Now an empty "master" field is ignored.
	- doveadm-server: Duplicate error lines were sent for failed commands.
	  This didn't normally cause visible problems, except when using
	  wildcards in usernames or -A parameter to go through multiple users.
	- doveadm-server: Logs written by doveadm-server were often missing log
	  prefixes, especially mail_log_prefix for mail commands. Logs sent to
	  doveadm TCP client were also missing log prefixes.
	- doveadm: v2.3 regression: batch command always crashes.
	- doveadm: v2.3.11 regression: Commands failed if ssl_cert or
	  ssl_key files weren't readable by the user running doveadm, even
	  though doveadm didn't actually use these settings
	- imap-hibernate: Process may crash at deinit:
	  Panic: file ioloop.c: line 928 (io_loop_destroy): assertion failed:
	  (ioloop->cur_ctx == NULL).
	- imap: Using imap_fetch_failure=no-after can cause assert-crash
	  with some IMAP commands if reading the mail fails (e.g. wrong cached
	  mail size). Fixes:
	  Panic: file index-mail-headers.c: line 198 (index_mail_parse_header_init):
	  assertion failed: (!mail->data.header_parser_initialized)
	- imap: v2.3.10 regression: When using INDEXPVT to enable private
	  \Seen flags (for shared or public namespaces) the STORE command did
	  not send untagged replies for the \Seen flag changes.
	- imap: v2.3.15 regression: If PREVIEW/SNIPPET is not the final FETCH
	  option in the command, the IMAP FETCH response is broken.
	- imap: v2.3.15 regression: MOVE command leaks mailbox if it can't be
	  opened and crashes at deinit:
	  Panic: file mail-user.c: line 229 (mail_user_deinit): assertion failed:
	  ((*user)->refcount == 1).
	- imapc: Copying nonexistent mail via imapc could have crashed. Fixes:
	  Panic: file mail-storage.c: line 2385 (mailbox_transaction_commit_get_changes):
	  assertion failed: (ret < 0 || seq_range_count(&changes_r->saved_uids) == save_count ||
	  array_count(&changes_r->saved_uids) == 0).
	- indexer: v2.3.15 regression: Process crashes if indexer-client
	  disconnects while it's waiting for command reply. This happened for
	  example if IMAP SEARCH triggered long fts indexing and the IMAP
	  client disconnected while waiting for the reply.
	- indexer: v2.3.15 regression: Process may have crashed in some situations.
	- indexer: v2.3.15 regression: indexer-worker processes may not have
	  reached the process_limit in some situations, possibly even using just
	  one indexer-worker process even though there were many indexing
	  requests queued.
	- lib-compression: Reading lz4 compressed mdbox mails may crash. Fixes:
	  Panic: file istream.c: line 345 (i_stream_read_memarea):
	  assertion failed: (!stream->blocking).
	- lib-compression: bench-compress crashes due to xz being read-only.
	- lib-lua: Fix linking libdict_lua for non-GNU linkers when Lua support
	  is disabled.
	- lib-mail: There was no limit on how large an email header name could be.
	  Processable header names are now limited to 1000 bytes.
	- lib-oauth2: Dovecot disallowed JWT tokens if their validity time was
	  older than token creation time (nbf < iat).
	- lib-storage: Reduce memory footprint of certain storage operations.
	- lib-storage: When listing mailboxes with storage name escape
	  characters (^ or .) as part of the mailbox name, the listing could
	  show corrupted mailbox names. Due to an issue in handling escaped
	  parent folders, the listing of other mailbox names would become
	  corrupted by prepending parts of the previously listed mailboxes
	  parent folder as prefix to the actual mailbox names. The corruption
	  can occur when using LAYOUT=INDEX and maildir or obox, or when using
	  the listescape plugin.
	- mail-crypt: Fix "-O" argument for "doveadm mailbox cryptokey password"
	  command to be a boolean, and not expect a string.
	- submission-login: Add support for not authenticating to next hop in
	  submission proxying.
	- submission-login: EHLO was not sent again after XCLIENT when doing

src/doveadm/doveadm-pw.c
/* Copyright (c) 2004-2018 Dovecot authors, see the included COPYING file */

#include "lib.h"
#include "array.h"
#include "password-scheme.h"
#include "randgen.h"
#include "doveadm.h"
#include "askpass.h"
#include "module-dir.h"

#include <ctype.h>
#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>

#define DEFAULT_SCHEME "CRYPT"

static struct module *modules = NULL;

static void cmd_pw(struct doveadm_cmd_context *cctx)
{
	const char *hash = NULL;
	const char *scheme = NULL;
	const char *plaintext = NULL;
	const char *test_hash = NULL;
	bool list_schemes = FALSE, reverse_verify = FALSE;
	int64_t rounds_int64;
	struct module_dir_load_settings mod_set;
	struct password_generate_params gen_params;
	i_zero(&gen_params);

	password_schemes_init();
	password_schemes_allow_weak(TRUE);

	i_zero(&mod_set);
	mod_set.abi_version = DOVECOT_ABI_VERSION;
	mod_set.require_init_funcs = TRUE;
	mod_set.ignore_dlopen_errors = TRUE;
	mod_set.debug = doveadm_debug;

	modules = module_dir_load_missing(modules, AUTH_MODULE_DIR, NULL, &mod_set);
	module_dir_init(modules);

	(void)doveadm_cmd_param_bool(cctx, "list", &list_schemes);
	(void)doveadm_cmd_param_str(cctx, "plaintext", &plaintext);
	if (doveadm_cmd_param_int64(cctx, "rounds", &rounds_int64)) {
		if (rounds_int64 > UINT_MAX)
			i_fatal("Invalid number of rounds: %"PRId64, rounds_int64);
		gen_params.rounds = rounds_int64;
	}
	(void)doveadm_cmd_param_str(cctx, "scheme", &scheme);
	if (doveadm_cmd_param_str(cctx, "test-hash", &test_hash))
		reverse_verify = TRUE;
	(void)doveadm_cmd_param_str(cctx, "user", &gen_params.user);
	(void)doveadm_cmd_param_bool(cctx, "reverse-verify", &reverse_verify);

	if (list_schemes) {
		ARRAY_TYPE(password_scheme_p) arr;
		const struct password_scheme *const *schemes;
		unsigned int i, count;
		t_array_init(&arr, 30);
		password_schemes_get(&arr);
		schemes = array_get(&arr, &count);
		for (i = 0; i < count; i++)
			printf("%s ", schemes[i]->name);
		printf("\n");
		module_dir_unload(&modules);
		password_schemes_deinit();
		return;
	}

	scheme = scheme == NULL ? DEFAULT_SCHEME : t_str_ucase(scheme);

	if (test_hash != NULL && plaintext == NULL)
		plaintext = t_askpass("Enter password to verify: ");
	while (plaintext == NULL) {
		const char *check;
		static int lives = 3;

		plaintext = t_askpass("Enter new password: ");
		check = t_askpass("Retype new password: ");
		if (strcmp(plaintext, check) != 0) {
			i_error("Passwords don't match!");
			if (--lives == 0)
				lib_exit(1);
			plaintext = NULL;
		}
	}

	if (!password_generate_encoded(plaintext, &gen_params, scheme, &hash))
		i_fatal("Unknown scheme: %s", scheme);
	if (reverse_verify) {
		const unsigned char *raw_password;
		size_t size;
		const char *error;

		if (test_hash != NULL) {
			scheme = password_get_scheme(&test_hash);
			if (scheme == NULL)
				i_fatal("Missing {scheme} prefix from hash");
			hash = test_hash;
		}

		if (password_decode(hash, scheme, &raw_password, &size,
				    &error) <= 0)
			i_fatal("reverse decode check failed: %s", error);

		if (password_verify(plaintext, &gen_params, scheme,
				    raw_password, size, &error) <= 0) {
			i_fatal("reverse password verification check failed: %s",
				error);
		}

		printf("{%s}%s (verified)\n", scheme, hash);
	} else {
		printf("{%s}%s\n", scheme, hash);
	}

	module_dir_unload(&modules);
	password_schemes_deinit();
}

struct doveadm_cmd_ver2 doveadm_cmd_pw = {
	.name = "pw",
	.cmd = cmd_pw,
	.usage = "[-l] [-p plaintext] [-r rounds] [-s scheme] [-t hash] [-u user] [-V]",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('l', "list", CMD_PARAM_BOOL, 0)
DOVEADM_CMD_PARAM('p', "plaintext", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('r', "rounds", CMD_PARAM_INT64, 0)
DOVEADM_CMD_PARAM('s', "scheme", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('t', "test-hash", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('u', "user", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('V', "reverse-verify", CMD_PARAM_BOOL, 0)
DOVEADM_CMD_PARAMS_END
};
